### PR TITLE
Remove config fields with CLI overrides using ~

### PR DESCRIPTION
**Pull request by RdoubleA**
## Problem
We provide an easy way to update or add config fields from a CLI override using the `key=value` syntax. However, we have no way of removing an existing field from a yaml config through the CLI. This was recently highlighted in #1028 with `text_completion_dataset` CLI examples in our tutorials. `text_completion_dataset` does not use the parameter `train_on_input`, but this is a default parameter set in all our configs that's supported by most of our datasets and is exposed intentionally to highlight this commonly used feature. If a user wants to quickly override our built-in configs with `text_completion_dataset`, they need to remove this parameter.

## Possible approaches
There are several ways we can approach this problem:
- **Remove `train_on_input` from all our configs**. `alpaca_dataset` already has this set to True by default, and so it is not needed. This will address our issue with `text_completion_dataset` (and other datasets that do not use this parameter) but we lose the visibility of this feature
- **Overriding a component clears all its kwargs by default**. Currently, when we override a component, like dataset with `dataset=torchtune.datasets.text_completion_dataset`, it keeps the previous kwargs in the config file. In this case, `train_on_input` remains, causing the error. We could instead delete all kwargs, requiring the users to specify them again in the CLI overrides, but this can get very tedious for components like the checkpointer which has many kwargs. Additionally, if we decide to enable nested components, this approach is just a no-go.
- **Ask users to just clone and modify**. Removing a field from a built-in config might be considered advanced behavior that should be handled directly in the config file. CLI overrides are meant to be a quick way to change fields but not a replacement for complex changes to configs. We could draw the line here.
- **Adding a way to remove params from CLI override**. This has been discussed numerous times offline as a solution to pain points in quick experimentation. The main downside with this is that it can make CLI overrides a bit intrusive and awkward

## Changes in this PR
- enable removing config fields in the YAML with CLI overrides using the `~` flag. This is inspired by how [Hydra](https://hydra.cc/docs/advanced/override_grammar/basic/#modifying-the-config-object) handles this. 
- remove `train_on_input` from all configs as only alpaca uses this and it defaults to true in the builder anyway
- Update all docs on text_completion_dataset, bug fixes from #1028   

#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)
	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ] 	- [ ]

### CODE REVIEW

File: docs/source/deep_dives/configs.rst [hunk 1]
```diff
@@ -236,6 +231,26 @@ If you would like to override a class or function in the config that is instanti
...
+    dataset=torchtune.datasets.slimorca_dataset dataset.train_on_input=True
+
+Removing config fields
+^^^^^^^^^^^^^^^^^^^^^^
+You may need to remove certain parameters from the config when changing components
+through overrides that require different keyword arguments. You can do so by using
+the `~` flag and specify the dotpath of the config field you would like to remove.
+For example, if you want to override a built-in config and use :func:`~torchtune.datasets.text_completion_dataset`
+which does not use the ``train_on_input`` parameter, you can delete it from the
+command-line.
```
Comment (ebsmothers) [inline docs/source/deep_dives/configs.rst:252]: Wait this is kinda no longer a relevant example, right?

File: docs/source/tutorials/datasets.rst [hunk 2]
```diff
@@ -184,24 +184,26 @@ files:
...
 
 .. code-block:: bash
 
     # Command line
-    tune run full_finetune_single_device --config llama3/8B_full_single_device \
-    dataset=torchtune.datasets.text_completion_dataset dataset.source=txt \
-    dataset.data_files=path/to/my_data.txt
+    tune run full_finetune_distributed --config llama3/8B_full \
+    dataset=torchtune.datasets.text_completion_dataset dataset.source=text \
+    dataset.data_files=path/to/my_data.txt dataset.split=train
```
Comment (ebsmothers) [inline docs/source/tutorials/datasets.rst:206]: Why this change? Shouldn't we have number of devices in here if we're doing this?

**Review by ebsmothers — APPROVED** [review]
Couple small comments but no major concerns. Also you may need to merge with latest main, there are a couple more configs under dev/ that will need to be updated as well

File: docs/source/tutorials/datasets.rst (hunk 2 above)
Comment (RdoubleA) [reply docs/source/tutorials/datasets.rst:206]: single device recipe requires bitsandbytes, which needs to be installed separately by the user. I'd rather not use that as an example. But yes, I need to add number of devices..
